bootstrap/kubeadm/internal/ignition/clc/clc.go

```diff
@@ -0,0 +1,192 @@
+	clct "github.com/flatcar-linux/container-linux-config-transpiler/config"
+	"github.com/pkg/errors"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/internal/cloudinit"
+)
+
+const (
+	clcTemplate = `---
+systemd:
+  units:
+    - name: kubeadm.service
+      enabled: true
+      contents: |
+        [Unit]
+        Description=kubeadm
+        [Service]
+        # To not restart the unit when it exits, as it is expected.
+        Type=oneshot
+        ExecStart=/etc/kubeadm.sh
+        [Install]
+        WantedBy=multi-user.target
+
+storage:
+  files:{{ range .WriteFiles }}
+  - path: {{.Path}}
+    {{ if ne .Permissions "" -}}
+    mode: {{.Permissions}}
+    {{ end -}}
+    contents:
+      inline: |
+{{.Content | Indent 8}}
+{{- end }}
+  - path: /etc/kubeadm.sh
+    mode: 0700
+    contents:
+      inline: |
+        #!/bin/bash
+        set -e
+
+        {{- range .PreKubeadmCommands }}
+        {{ . }}
+        {{- end }}
+
+        {{ .KubeadmCommand }}
+        rm /etc/kubeadm.yml
+
+        {{- range .PostKubeadmCommands }}
+        {{ . }}
+        {{- end }}
+  - path: /etc/kubeadm.yml
+    mode: 0600
+    contents:
+      inline: |
+        ---
+{{ .KubeadmConfig | Indent 8 }}
+`
+)
+
+type render struct {
+	*cloudinit.BaseUserData
+
+	KubeadmConfig string
+}
+
+func defaultTemplateFuncMap() template.FuncMap {
+	return template.FuncMap{
+		"Indent": templateYAMLIndent,
+	}
+}
+
+func templateYAMLIndent(i int, input string) string {
+	split := strings.Split(input, "\n")
+	ident := "\n" + strings.Repeat(" ", i)
+	return strings.Repeat(" ", i) + strings.Join(split, ident)
+}
+
+func renderCLC(input *cloudinit.BaseUserData, kubeadmConfig string) ([]byte, error) {
+	if input == nil {
+		return nil, errors.New("empty base user data")
+	}
+
+	t := template.Must(template.New("template").Funcs(defaultTemplateFuncMap()).Parse(clcTemplate))
+
+	data := render{
+		BaseUserData:  input,
+		KubeadmConfig: kubeadmConfig,
+	}
+
+	var out bytes.Buffer
+	if err := t.Execute(&out, data); err != nil {
+		return nil, errors.Wrapf(err, "failed to render template")
+	}
+
+	return out.Bytes(), nil
+}
+
+func Render(input *cloudinit.BaseUserData, clc *bootstrapv1.ContainerLinuxConfig, kubeadmConfig string) ([]byte, string, error) {
+	if clc == nil {
+		return nil, "", errors.New("get empty CLC config")
+	}
+
+	clcBytes, err := renderCLC(input, kubeadmConfig)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "rendering CLC configuration")
+	}
+
+	userData, warnings, err := buildIgnitionConfig(clcBytes, clc)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "building Ignition config")
+	}
+
+	return userData, warnings, nil
+}
+
+func buildIgnitionConfig(baseCLC []byte, clc *bootstrapv1.ContainerLinuxConfig) ([]byte, string, error) {
+	// We control baseCLC config, so treat it as strict.
+	ign, warnings, err := clcToIgnition(baseCLC, true)
```

Comment:
Since we're setting `strict`, the second return value can be ignored here.

Reply:
True, fixed, thanks!